engines/qemu/network/iptables.go

```diff
@@ -72,35 +74,66 @@ func ipTableRules(tapDevice string, ipPrefix string, delete bool) [][]string {
 		{"-j", "REJECT", "--reject-with", "icmp-net-prohibited"},
 	})
 
+	// Create VPN forwarding rules
+	forwardVPNInputRules := [][]string{}  // Will be prepended fwd_input_...
+	forwardVPNOutputRules := [][]string{} // Will be prepended fwd_output_...
+	for _, vpn := range vpns {
+		for _, ip := range vpn.Routes() {
+			ipv4 := ip.To4()
+			if ipv4 == nil {
+				debug("Skipping IPv6 route to VPN: %s", ip.String())
+				continue // Skip IPv6 for now
+			}
+			route := ipv4.String()
+			// Allow tap device -> VPN, if source subnet and target tap device matches
+			forwardVPNInputRules = append(forwardVPNInputRules, []string{
+				"-d", route, "-o", vpn.DeviceName(), "-s", subnet, "-j", "ACCEPT",
+			})
+			// Allow VPN -> tap device, if destination subnet matches and connection
+			// is already established.
+			forwardVPNOutputRules = append(forwardVPNOutputRules, []string{
+				"-s", route, "-i", vpn.DeviceName(), "-d", subnet, "-m", "state", "--state", "RELATED,ESTABLISHED", "-j", "ACCEPT",
+			})
+		}
+	}
+
 	// Rules for filtering FORWARD from this tap device
-	forwardInputRules := prefixCommands([]string{"iptables", "-w", xtableLockWait, ruleAction, "fwd_input_" + tapDevice}, [][]string{
-		// Reject out-going from tctap1 to private subnets
-		{"-d", "10.0.0.0/8", "-j", "REJECT", "--reject-with", "icmp-net-unreachable"},
-		{"-d", "172.16.0.0/12", "-j", "REJECT", "--reject-with", "icmp-net-unreachable"},
-		{"-d", "169.254.0.0/16", "-j", "REJECT", "--reject-with", "icmp-net-unreachable"},
-		{"-d", "192.168.0.0/16", "-j", "REJECT", "--reject-with", "icmp-net-unreachable"},
-		// Allow out-going from tctap1 with correct source subnet
-		{"-o", "eth0", "-s", subnet, "-j", "ACCEPT"},
-		// Allow tctap1 -> tctap1 within allowed subnet
-		{"-o", "tctap1", "-s", subnet, "-j", "ACCEPT"},
-		// Reject all other input for forwarding from tap-device
-		{"-j", "REJECT", "--reject-with", "icmp-net-prohibited"},
-	})
+	forwardInputRules := prefixCommands([]string{"iptables", "-w", xtableLockWait, ruleAction, "fwd_input_" + tapDevice}, append(
+		// Allow tap device -> VPN
+		forwardVPNInputRules,
+		[][]string{
+			// Reject out-going from this tap device to private subnets
```

Comment:
Cool. Makes sense. Let's add that as a comment here then. I think this section could use really heavy commenting as to the motivation behind each rule.